File: backend/app/models/campus_hostel.py
"""Campus hostel: blocks, rooms, allocations, out-passes and visitors."""

from sqlalchemy import (
    Boolean,
    Column,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    UniqueConstraint,
    func,
)

from app.core.database import Base
from app.models.tuition import MONEY


ROOM_TYPES = ("single", "double", "triple", "dormitory")
PASS_STATUSES = ("pending", "approved", "rejected", "returned")


class CampusHostelBlock(Base):
    __tablename__ = "campus_hostel_blocks"
    id = Column(Integer, primary_key=True)
    institution_id = Column(Integer, ForeignKey("institutions.id"), nullable=False, index=True)
    name = Column(String(120), nullable=False)
    warden_member_id = Column(Integer, ForeignKey("institution_members.id"), nullable=True)
    gender = Column(String(10), nullable=False, default="any")
    fee_amount = Column(MONEY, nullable=True)
    currency = Column(String(3), nullable=False, default="INR")
    fee_plan_id = Column(Integer, ForeignKey("tuition_fee_plans.id"), nullable=True)
    active = Column(Boolean, nullable=False, default=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    __table_args__ = (UniqueConstraint("institution_id", "name", name="uq_campus_hostel_block"),)


class CampusHostelRoom(Base):
    __tablename__ = "campus_hostel_rooms"
    id = Column(Integer, primary_key=True)
    block_id = Column(Integer, ForeignKey("campus_hostel_blocks.id"), nullable=False, index=True)
    number = Column(String(20), nullable=False)
    floor = Column(String(20), nullable=False, default="")
    room_type = Column(String(20), nullable=False, default="double")
    capacity = Column(Integer, nullable=False, default=2)
    active = Column(Boolean, nullable=False, default=True)
    __table_args__ = (UniqueConstraint("block_id", "number", name="uq_campus_hostel_room"),)


class CampusHostelAllocation(Base):
    __tablename__ = "campus_hostel_allocations"
    id = Column(Integer, primary_key=True)
    institution_id = Column(Integer, ForeignKey("institutions.id"), nullable=False, index=True)
    room_id = Column(Integer, ForeignKey("campus_hostel_rooms.id"), nullable=False, index=True)
    member_id = Column(Integer, ForeignKey("institution_members.id"), nullable=False)
    fee_assignment_id = Column(Integer, ForeignKey("tuition_fee_assignments.id"), nullable=True)
    status = Column(String(20), nullable=False, default="active")
    checked_in_on = Column(Date, nullable=False)
    checked_out_on = Column(Date, nullable=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    __table_args__ = (Index("ix_campus_hostel_allocations_member_status", "member_id", "status"),)


class CampusHostelPass(Base):
    __tablename__ = "campus_hostel_passes"
    id = Column(Integer, primary_key=True)
    institution_id = Column(Integer, ForeignKey("institutions.id"), nullable=False, index=True)
    member_id = Column(Integer, ForeignKey("institution_members.id"), nullable=False, index=True)
    kind = Column(String(20), nullable=False, default="outpass")
    reason = Column(String(500), nullable=False, default="")
    leaves_at = Column(DateTime(timezone=True), nullable=False)
    returns_at = Column(DateTime(timezone=True), nullable=False)
    status = Column(String(20), nullable=False, default="pending", index=True)
    decided_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    decided_at = Column(DateTime(timezone=True), nullable=True)
    decision_note = Column(String(500), nullable=False, default="")
    returned_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())


class CampusHostelVisitor(Base):
    __tablename__ = "campus_hostel_visitors"
    id = Column(Integer, primary_key=True)
    institution_id = Column(Integer, ForeignKey("institutions.id"), nullable=False, index=True)
    member_id = Column(Integer, ForeignKey("institution_members.id"), nullable=False, index=True)
    visitor_name = Column(String(120), nullable=False)
    relation = Column(String(60), nullable=False, default="")
    phone = Column(String(20), nullable=False, default="")
    checked_in_at = Column(DateTime(timezone=True), nullable=False, index=True)
    checked_out_at = Column(DateTime(timezone=True), nullable=True)
    recorded_by = Column(Integer, ForeignKey("users.id"), nullable=False)

File: frontend/src/api/campus-hostel.ts
import { api } from "./axios";

const root = (id: number) => `/institutions/${id}/hostel`;

export type RoomType = "single" | "double" | "triple" | "dormitory";
export type PassStatus = "pending" | "approved" | "rejected" | "returned";

export interface HostelRoom {
  id: number;
  block_id: number;
  number: string;
  floor: string;
  room_type: RoomType;
  capacity: number;
  occupied: number;
  active: boolean;
}

export interface HostelBlock {
  id: number;
  name: string;
  warden_member_id: number | null;
  warden_name: string | null;
  gender: "any" | "male" | "female";
  fee_amount: number | null;
  currency: string;
  fee_plan_id: number | null;
  active: boolean;
  capacity: number;
  occupied: number;
  rooms: HostelRoom[];
}

export interface RoomInput {
  number: string;
  floor: string;
  room_type: RoomType;
  capacity: number;
  active: boolean;
}

export interface BlockInput {
  name: string;
  warden_member_id: number | null;
  gender: "any" | "male" | "female";
  fee_amount: string | null;
  currency: string;
  rooms: RoomInput[];
}

export interface HostelAllocation {
  id: number;
  room_id: number;
  room_number: string | null;
  block_id: number | null;
  block_name: string | null;
  member_id: number;
  student_name: string | null;
  fee_assignment_id: number | null;
  status: "active" | "ended";
  checked_in_on: string;
  checked_out_on: string | null;
}

export interface HostelOccupancy {
  blocks: HostelBlock[];
  allocations: HostelAllocation[];
  capacity: number;
  occupied: number;
}

export interface HostelPass {
  id: number;
  member_id: number;
  student_name: string | null;
  kind: "outpass" | "leave";
  reason: string;
  leaves_at: string;
  returns_at: string;
  status: PassStatus;
  decided_by: number | null;
  decided_at: string | null;
  decision_note: string;
  returned_at: string | null;
  created_at: string | null;
}

export interface HostelVisitor {
  id: number;
  member_id: number;
  student_name: string | null;
  visitor_name: string;
  relation: string;
  phone: string;
  checked_in_at: string;
  checked_out_at: string | null;
}

export interface HostelMe {
  resident: boolean;
  allocation: HostelAllocation | null;
  passes: HostelPass[];
}

export const hostelApi = {
  blocks: async (id: number) => (await api.get<HostelBlock[]>(`${root(id)}/blocks`)).data,
  createBlock: async (id: number, data: BlockInput) => (await api.post<HostelBlock>(`${root(id)}/blocks`, data)).data,
  updateBlock: async (id: number, blockId: number, data: Partial<Pick<BlockInput, "name" | "warden_member_id" | "gender">> & { active?: boolean }) =>
    (await api.patch<HostelBlock>(`${root(id)}/blocks/${blockId}`, data)).data,
  saveRooms: async (id: number, blockId: number, rooms: RoomInput[]) =>
    (await api.put<HostelBlock>(`${root(id)}/blocks/${blockId}/rooms`, { rooms })).data,
  occupancy: async (id: number) => (await api.get<HostelOccupancy>(`${root(id)}/occupancy`)).data,
  occupancyCsv: async (id: number) => (await api.get<Blob>(`${root(id)}/occupancy.csv`, { responseType: "blob" })).data,
  allocate: async (id: number, roomId: number, memberId: number) =>
    (await api.post<HostelAllocation>(`${root(id)}/rooms/${roomId}/allocations`, { member_id: memberId })).data,
  checkout: async (id: number, allocationId: number) =>
    (await api.post<HostelAllocation>(`${root(id)}/allocations/${allocationId}/checkout`)).data,
  passes: async (id: number, params: { status?: PassStatus; student_user_id?: number } = {}) =>
    (await api.get<HostelPass[]>(`${root(id)}/passes`, { params })).data,
  createPass: async (id: number, data: { kind: "outpass" | "leave"; reason: string; leaves_at: string; returns_at: string }) =>
    (await api.post<HostelPass>(`${root(id)}/passes`, data)).data,
  approvePass: async (id: number, passId: number, note = "") =>
    (await api.post<HostelPass>(`${root(id)}/passes/${passId}/approve`, { note })).data,
  rejectPass: async (id: number, passId: number, note: string) =>
    (await api.post<HostelPass>(`${root(id)}/passes/${passId}/reject`, { note })).data,
  returnPass: async (id: number, passId: number) => (await api.post<HostelPass>(`${root(id)}/passes/${passId}/return`)).data,
  visitors: async (id: number, day?: string) => (await api.get<HostelVisitor[]>(`${root(id)}/visitors`, { params: { day } })).data,
  logVisitor: async (id: number, data: { member_id: number; visitor_name: string; relation: string; phone: string }) =>
    (await api.post<HostelVisitor>(`${root(id)}/visitors`, data)).data,
  visitorCheckout: async (id: number, visitorId: number) =>
    (await api.post<HostelVisitor>(`${root(id)}/visitors/${visitorId}/checkout`)).data,
  me: async (id: number, studentUserId?: number) =>
    (await api.get<HostelMe>(`${root(id)}/me`, { params: { student_user_id: studentUserId } })).data,
};
